### ISSUE TITLE
Command '$eval' failed: not authorized on <myDB> to execute command 

### ISSUE THREAD

**Issue by mridulsri** [sergun/Hangfire.Mongo#4]
when the MongoDB running under --auth mode. The below exception throwing in  MongoExtensions helper class. 
Error:
Command '$eval' failed: not authorized on hangfire_db to execute command 
{ $eval: new Date() } 
(response: { "ok" : 0.0, "errmsg" : "not authorized on hangfire_db to execute command { $eval: new Date() }", "code" : 13 })

I can able to connect properly using mongodb credentials. But not able to execute below code.

``` public static DateTime GetServerTimeUtc(this MongoDatabase database)
 {
       return database.Eval(new EvalArgs
                {
                    Code = new BsonJavaScript("new Date()")
                }).ToUniversalTime();
 }```

For now I have a workaround for same. 
public static DateTime GetServerTimeUtc(this MongoDatabase database)
        {
            try
            {
                return database.Eval(new EvalArgs
                {
                    Code = new BsonJavaScript("new Date()")
                }).ToUniversalTime();
            }
            catch
            {
                String dateTimeString = DateTime.UtcNow.ToShortDateString();
                DateTime baseLined = Convert.ToDateTime(dateTimeString);
                return baseLined.ToUniversalTime();
            }
        }
This is working fine with mongodb in --auth mode.
But more then one server adding in development machine db.

**Comment by sergun** [comment]
Hi @mridulsri 
Maybe some suggestions how to handle this?

**Comment by mridulsri** [comment]
Hi Sergey,

As per MongoDB official site the behavior of "eval" commond is different
form other one.
It takes write -lock, restriction with Sharded Data and access-control. You
can find more details on MongoDB Official site(
http://docs.mongodb.org/manual/reference/command/eval/)
Since our MongoDB Server running in "auth" mode, So I can not use or
execute the "eval" command. Because of some security concern.

So,I found one solution for me. I have modified the MongoExtensions class
method "GetServerTimeUtc". Below is the definition which I have Modified.

public static DateTime GetServerTimeUtc(this MongoDatabase database)
{
try
{
//This try block execute properly if MongoDB server not running in "auth"
mode, otherwise it throw exception. Which I have handle in catch block and
return server time.
return database.Eval(new EvalArgs
{
Code = new BsonJavaScript("new Date()")
}).ToUniversalTime();
}
catch
{
//if MongoDB server running in "auth" mode.
String dateTimeString = DateTime.UtcNow.ToShortDateString();
DateTime baseLined = Convert.ToDateTime(dateTimeString);
return baseLined.ToUniversalTime();
}
}

Thanking You.
Happy coding.

Regards
Mridul Srivastava
(India)

On Thu, Jul 2, 2015 at 6:24 AM, Sergey Zwezdin <notifications@github.com>
wrote:

> Hi @mridulsri <https://github.com/mridulsri>
> Maybe some suggestions how to handle this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sergun/Hangfire.Mongo/issues/4#issuecomment-117864854>
> .
>



-- 
Thanks & Regards
Mridul Srivastava
New Delhi
Mob. #9350272167
eMail:- mridul_gzp@yahoo.com, mridul81@gmail.com
